### PR TITLE
Automatic par

### PR DESCRIPTION
**Pull request by tissue3**
Ugly code. Welcome to make it better.

### CODE REVIEW

**Comment by sgpthomas** [comment]
@tissue3 Cleaned up the code. Let me know what you think

File: calyx/src/passes/automatic_par.rs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            None => return Ok(Action::Continue),
+        };
+
+        // vec of new control statements including all elements up to the first
+        // enable that we find
+        let mut new_stmts: Vec<ast::Control> = seq.stmts[..start].to_vec();
+
+        // start interation from the second item because the first
+        // is in `cmp_stmt`
+        for stmt in seq.stmts[start + 1..].iter() {
```
Comment (rachitnigam) [inline calyx/src/passes/automatic_par.rs]: Correct me if I understood this code wrong. It is trying to the do the following:
```
(seq
  (enable A)   <-- start
  (enable B)    ---
  (enable C)        | <-- remaining
   . . .        ---
)
```

For every `start` in the sequence, iterate over every other component, detect if they have any components in common. If they don't, parallelize them.

From the loops below, it is hard to parse this intention. Plus, there is a bunch of unnecessary re-computation being done with this pass. When we start checking `B`, we'll have to recompute the components in `C` and so on. I recommend two things:

1. Document the crap out of this pass. Use some [ASCII art](http://asciiflow.com/) and really nail down what this pass should be doing.
2. Consider splitting this into two passes. One that computes all the components on the inputs and outputs of each port, and another one that simply checks this data structure for conflicts. This can also be a good test to see how pass composition works.

Comment (sgpthomas) [reply calyx/src/passes/automatic_par.rs:53]: There's no looping over `start`. `start` just represents the first enable we find. This is just a small optimization to slightly reduce the statements we check.

Doing this in two passes would be less efficient because here we are accumulating enable statements as we go. To show what I mean, suppose
```
(enable A)
(enable B)
(enable C)
```
is equivalent to
```
(enable A B C)
```
In this pass we check if `(enable A)` can be merged with `(enable B)`. In this case they can be so we make the statement `(enable A B)`. Then we check this new statement against `(enable C)` to see if it can be merged. In this case it can, so we produce `(enable A B C)`. etc.

If we did the analysis first, you would have to check all enable statements pairwise, because (using ~ for 'safe to merge with') `(enable A) ~ (enable B) && (enable B) ~ (enable C)` does not imply `(enable A) ~ (enable C)`. To see this imagine 3 components with only an edge between `A` and `C`.

**Comment by rachitnigam** [comment]
I see. I've added more documentation for the code and refactored it a bit more. Please the code for `has_conflicts` and make sure it works. 

Also, add more tests to document the behavior for what happens when there is a non-mergable node in the middle of two mergeable nodes.

File: calyx/src/passes/automatic_par.rs [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+                        new_stmts.push(Control::enable(cmp_acc.comps.clone()));
+                        // there was a conflict, start, a new accumulator.
+                        cmp_acc = enables2.clone();
+                    } else {
+                        // If there is no conflict, update the current
+                        // component accumulator.
+                        cmp_acc.comps.append(&mut enables2.comps.clone());
+                    }
+                }
+                _ => new_stmts.push(stmt.clone()),
```
Comment (sgpthomas) [inline calyx/src/passes/automatic_par.rs:147]: I realized this is not quite correct. We also have to change the accumulator. Otherwise something like
```
(seq (enable A B)
      (seq ...)
      (enable C D))
```
would get collapsed to
```
(seq (seq ...)
      (enable A B C D))
```

Comment (rachitnigam) [reply calyx/src/passes/automatic_par.rs:147]: Ah, you’re right. Add a relevant test and feel free to fix if you get to it before me!

**Comment by sgpthomas** [comment]
This pass doesn't correctly deal with the case where subsequent enables have components in common.
For example,
```
(enable A B)
(enable B C)
```
might be collapsed if there are no edges between these two subgraphs.
